src/firefly/java/edu/caltech/ipac/firefly/server/visualize/fitseval/SpectralCubeEval.java

```diff
@@ -51,37 +49,109 @@ public static int findWaveTabHDU(BasicHDU[] HDUs) {
         return tabHduIdx;
     }
 
+
+    /**
+     * If a FITS file contains multiple XTENSION HDUs (headerdata
+     * units) with the specified EXTNAME, EXTLEVEL, and
+     * EXTVER, then the result of the WCS table lookup is undefined.
+     * If the specified FITS BINTABLE contains no column, or multiple
+     * columns, with the specified TTYPEn, then the result of the
+     * WCS table lookup is undefined. The specified FITS BINTABLE
+     * must contain only one row.
+     * No units conversions are to
+     * @return
+     */
+    private static boolean isValidWCSForTab(BasicHDU[] HDUs){
+
+        List<String> extNames = new ArrayList<String>();
+        List<Integer> extLevels= new ArrayList<Integer>();
+        List<Integer> extVers  = new ArrayList<Integer>();
+
+
+        for (int i=0; i<HDUs.length; i++){
+            Header header= HDUs[i].getHeader();
+            String xtension= header.getStringValue("XTENSION");
+            if ("BINTABLE".equals(xtension)) {
+                extNames.add(header.getStringValue("EXTNAME"));
+                extLevels.add( header.getIntValue("EXTLEVEL"));
+                extVers.add( header.getIntValue("EXTVER"));
+
+            }
+        }
+
+
+        for (int i=0; i<extNames.size(); i++){
+            for (int j=i+1; j<extNames.size();j++){
+                if (extNames.get(i) == extNames.get(j) &&
+                   extLevels.get(i) == extLevels.get(j) &&
+                   extVers.get(i)   == extVers.get(j) ){
+                    return false;
+                }
+            }
+        }
+
+
+        return true;
+    }
     /**
      * NOTE:  HDUs here is the array of all HDUs in the FITs file.
-     * For each image HDU, look for its coressonding BinaryTableHDU.  This only applies for WAVE-TAB case
+     *        For each image HDU, look for its corresponding BinaryTableHDU.  This only applies for WAVE-TAB case.
+     *        The -TAB implementation is more complicated than most other WCS conventions because the coordinate system
+     *        is not completely defined by keywords in a single FITS header.
+     *        The necessary WCS parameters that are in general distributed over two FITS HDUs and in the body
+     *        of the WCS extension table.
      * @param HDUs
      * @param hduIdx
      * @return
      */
     public static int findWaveTabHDU(BasicHDU[] HDUs, int hduIdx) {
 
         if (HDUs.length<2) return -1;
+        if (!isValidWCSForTab(HDUs)) return -1;
+
         BasicHDU hdu= HDUs[hduIdx];
         Header header= hdu.getHeader();
         String xtensionPrim= header.getStringValue("XTENSION");
         if (xtensionPrim!=null && xtensionPrim.equals("BINTABLE")) return -1;
         String ctype= header.getStringValue("CTYPE3");
         if (!"WAVE-TAB".equals(ctype)) return -1;
         String waveHDUName= header.getStringValue("PS3_0");
```

Comment:
I think you should do this section before calling isValidWCSForTab

Reply:
Good suggestion!  Thanks, I will change it.

Reply:
@robyww  I changed the order of the validation as you suggested.  I did not pay attention about the efficiency issue about the existing codes. Thanks for pointing it out.